Workflow versioning

websupporter opened the pull request:
Fixes [MAILPOET-4430]

[MAILPOET-4430]: https://mailpoet.atlassian.net/browse/MAILPOET-4430?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

websupporter committed:
Update database entry only when an actual change happend

[MAILPOET-4430]

websupporter committed:
Separate workflow versionable data into workflow_version table

[MAILPOET-4430]

websupporter committed:
Add version_id to Workflows

[MAILPOET-4430]

websupporter committed:
Add version_id to WorkflowRun

[MAILPOET-4430]

websupporter committed:
Query workflow by version in StepRunner

[MAILPOET-4430]

websupporter committed:
Test WorkflowStorage

[MAILPOET-4430]

websupporter committed:
Add created and updated columns to versions table

[MAILPOET-4430]

websupporter committed:
Introduce Workflow::equals() and move decision to update a record to storage

[MAILPOET-4430]

websupporter committed:
Update updatedAt when values are changed

[MAILPOET-4430]

JanJakes commented on mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php:
```diff
@@ -10,43 +10,80 @@
+      version.id DESC
+LIMIT
+      0,1; ",
+      $id
```
Could we indent this similarly to normal PHP code (no left out-dent, sub-indenting by 2 spaces)? We seem to do that [in existing code](https://github.com/mailpoet/mailpoet/blob/c8278dde2a066f1f4602ef32151cc8d4ad22ca90/mailpoet/lib/Newsletter/NewslettersRepository.php#L268) and I guess it may look a tad bit nicer.

This applies also to other similar occurrences in this PR.

websupporter replied:
Let's keep it consistent :+1:
Addressed in [f937c5d](https://github.com/mailpoet/mailpoet/pull/4289/commits/f937c5d5c190c440ec2ffc7ba55bf65ee0296ce8)

JanJakes commented on mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php:
```diff
@@ -10,43 +10,80 @@
+WHERE
+      version.id = (SELECT Max(id) FROM $versionTable WHERE workflow_id= version.workflow_id)
+ORDER BY
+      workflow.id DESC; ";
```
I was checking where we're using this, and it seems to be only [the `WorkflowsGetEndpoint`](https://github.com/mailpoet/mailpoet/blob/c8278dde2a066f1f4602ef32151cc8d4ad22ca90/mailpoet/lib/Automation/Engine/Endpoints/Workflows/WorkflowsGetEndpoint.php). Considering it's a "listing" endpoint and [seeing the data it uses now](https://github.com/mailpoet/mailpoet/blob/c8278dde2a066f1f4602ef32151cc8d4ad22ca90/mailpoet/lib/Automation/Engine/Endpoints/Workflows/WorkflowsGetEndpoint.php#L29), I suppose we could just keep this method touch only the `workflows` table, and maybe rename it to `getWorkflowsList` to indicate it doesn't return full workflows, only the listing data.

websupporter replied:
I see why it would be beneficial to do so. On the other side, we switch explicitly from `LEFT JOIN` to `INNER JOIN` in the other queries to prevent invalid records being returned. And I think we should follow this thought here too. What do you think?

JanJakes replied:
@websupporter Ahh, right, we still need to use the `Workflow` data type. OK, let's keep it consistent. We can optimize it later by adding a read-only dedicated data type, such as `WorkflowListRecord`.

JanJakes commented on mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php:
```diff
@@ -73,9 +121,24 @@ public function getActiveTriggerKeys(): array {
+      version.trigger_keys
+FROM
+     $workflowTable AS workflow
+        LEFT JOIN
```
This is a small detail, but we may use just `JOIN` (inner join, I mean) as workflow without a record in `workflow_versions` would be invalid.

websupporter replied:
Addressed in [2ebd723](https://github.com/mailpoet/mailpoet/pull/4289/commits/2ebd7236d596787009393ed244de12682ab52e40)

JanJakes commented on mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php:
```diff
@@ -85,4 +148,24 @@ public function getActiveWorkflowsByTrigger(Trigger $trigger): array {
     }, (array)$data);
   }
+
+  private function workflowHeaderData(Workflow $workflow): array {
```
To satisfy our naming conventions, I guess it needs a verb, such as `getWorkflowHeaderData`.

websupporter replied:
Addressed in [f8785cd](https://github.com/mailpoet/mailpoet/pull/4289/commits/f8785cd3ad72147ab76f556977f32919d2619ea2)

JanJakes commented on mailpoet/lib/Automation/Engine/Migrations/Migrator.php:
```diff
@@ -47,6 +47,7 @@ public function createSchema(): void {
       CREATE TABLE {$this->prefix}workflow_runs (
         id int(11) unsigned NOT NULL AUTO_INCREMENT,
         workflow_id int(11) unsigned NOT NULL,
+        version_id int(11) unsigned NOT NULL,
```
Do we now need to store both `version_id` and `workflow_id`? Or... should we?
I guess in the `WorkflowRun` we can have both, but on the DB level, I'm not sure if it's needed. On the other hand, there's no harm saving both... or is there? 🤔  What do you think?

websupporter replied:
Technically speaking the `workflow_id` would not be necessary. You can join the version data base with the unique `version_id` and conclude what is the `workflow_id`. But to me, it is a different table. Maybe I just want quickly to fetch `Select workflow.id, workflow.name, count(workflow_runs.id) as num from workflow, workflow_runs where workflow_runs.workflow_id = workflow.id` or similar queries, where I really do not need data from the version table at all. This column would enable me to do so. Also, in debugging scenarios, I think it helps, to have quickly a high level information what workflow a run belongs to.

The harm I see is in more bytes stored, but I think it is convenient information to have in this table.

JanJakes replied:
@websupporter Makes sense! Let's keep it.

JanJakes reviewed (commented):
Great stuff! I have just a few comments/questions, nothing substantial.

websupporter committed:
Rename workflowHeaderData method

[MAILPOET-4430]

websupporter committed:
Use INNER JOIN to combine workflows with their versions

[MAILPOET-4430]

websupporter committed:
Fix intendation for SQL strings

[MAILPOET-4430]

JanJakes reviewed (commented):
LGTM! There are some conflicts now, unfortunately. Could you rebase, @websupporter? Thanks!

websupporter committed:
Fix use statements in test

[MAILPOET-4430]

websupporter committed:
Remove outdated $change variable

The usage of $change was a leftover from a rebase which was
necessary.

[MAILPOET-4430]

websupporter committed:
Query a workflow in a specific version in the StepHandler

The StepHandler must run a specific version of a workflow. Therefore, we need to
query this specific Workflow.

[MAILPOET-4430]

websupporter committed:
Use correct error code and translator comment

[MAILPOET-4430]

websupporter commented:
@JanJakes thanks for the review. Rebased. Please have a brief look again, as it was a bigger rebase, but I think I got everything. :bow:
